File: backend/app/schemas/ecosystem.py
from datetime import datetime

from pydantic import BaseModel

# ─── EcosystemProject ─────────────────────────────────────────────────────────

class ProjectCreate(BaseModel):
    name: str
    platform: str           # github / gitee / gitcode
    org_name: str
    repo_name: str | None = None
    community_id: int | None = None
    description: str | None = None
    tags: list[str] = []
    auto_sync_enabled: bool = True
    sync_interval_hours: int | None = None   # None = 使用全局默认值


class ProjectUpdate(BaseModel):
    name: str | None = None
    platform: str | None = None
    org_name: str | None = None
    repo_name: str | None = None
    community_id: int | None = None
    description: str | None = None
    tags: list[str] | None = None
    is_active: bool | None = None
    auto_sync_enabled: bool | None = None
    sync_interval_hours: int | None = None


class ProjectListOut(BaseModel):
    id: int
    name: str
    platform: str
    org_name: str
    repo_name: str | None
    community_id: int | None
    description: str | None
    tags: list[str]
    is_active: bool
    last_synced_at: datetime | None
    created_at: datetime
    auto_sync_enabled: bool
    sync_interval_hours: int | None

    model_config = {"from_attributes": True}


class ProjectOut(ProjectListOut):
    description: str | None
    tags: list[str]
    added_by_id: int | None

    model_config = {"from_attributes": True}


# ─── EcosystemContributor ─────────────────────────────────────────────────────

class ContributorOut(BaseModel):
    id: int
    project_id: int
    github_handle: str
    display_name: str | None
    avatar_url: str | None
    role: str | None
    commit_count_90d: int | None
    pr_count_90d: int | None
    review_count_90d: int | None
    star_count: int | None
    followers: int | None
    company: str | None
    location: str | None
    first_contributed_at: datetime | None
    person_id: int | None
    last_synced_at: datetime

    model_config = {"from_attributes": True}


class PaginatedContributors(BaseModel):
    items: list[ContributorOut]
    total: int
    page: int
    page_size: int


# ─── Sync Result ──────────────────────────────────────────────────────────────

class SyncResult(BaseModel):
    created: int
    updated: int
    errors: int

File: frontend/src/api/ecosystem.ts
import apiClient from './index'

export interface EcosystemProject {
  id: number
  name: string
  platform: string
  org_name: string
  repo_name: string | null
  community_id: number | null
  description: string | null
  tags: string[]
  is_active: boolean
  last_synced_at: string | null
  added_by_id: number | null
  created_at: string
  auto_sync_enabled: boolean
  sync_interval_hours: number | null
}

export interface EcosystemContributor {
  id: number
  project_id: number
  github_handle: string
  display_name: string | null
  avatar_url: string | null
  role: string | null
  commit_count_90d: number | null
  pr_count_90d: number | null
  star_count: number | null
  followers: number | null
  person_id: number | null
  last_synced_at: string
}

export interface PaginatedContributors {
  items: EcosystemContributor[]
  total: number
  page: number
  page_size: number
}

export interface SyncResult {
  created: number
  updated: number
  errors: number
}

export interface ProjectCreateData {
  name: string
  platform: string
  org_name: string
  repo_name?: string
  community_id?: number | null
  description?: string
  tags?: string[]
  auto_sync_enabled?: boolean
  sync_interval_hours?: number | null
}

export interface ProjectUpdateData {
  name?: string
  platform?: string
  org_name?: string
  repo_name?: string | null
  community_id?: number | null
  description?: string
  tags?: string[]
  is_active?: boolean
  auto_sync_enabled?: boolean
  sync_interval_hours?: number | null
}

export const listProjects = () =>
  apiClient.get<EcosystemProject[]>('/ecosystem').then(r => r.data)

export const createProject = (data: ProjectCreateData) =>
  apiClient.post<EcosystemProject>('/ecosystem', data).then(r => r.data)

export const getProject = (pid: number) =>
  apiClient.get<EcosystemProject>(`/ecosystem/${pid}`).then(r => r.data)

export const updateProject = (pid: number, data: ProjectUpdateData) =>
  apiClient.patch<EcosystemProject>(`/ecosystem/${pid}`, data).then(r => r.data)

export const syncProject = (pid: number) =>
  apiClient.post<SyncResult>(`/ecosystem/${pid}/sync`).then(r => r.data)

export const listContributors = (pid: number, params?: {
  q?: string
  unlinked?: boolean
  page?: number
  page_size?: number
}) =>
  apiClient.get<PaginatedContributors>(`/ecosystem/${pid}/contributors`, { params }).then(r => r.data)

export const importContributorToPeople = (pid: number, handle: string) =>
  apiClient.post<{ action: string; person_id: number }>(`/ecosystem/${pid}/contributors/${handle}/import-person`).then(r => r.data)

export const deleteProject = (pid: number) =>
  apiClient.delete(`/ecosystem/${pid}`)
